cnf-tests/testsuites/e2esuite/ovs_qos/ovs_qos.go

```diff
@@ -0,0 +1,644 @@
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+	"github.com/openshift-kni/cnf-features-deploy/cnf-tests/testsuites/pkg/client"
+	"github.com/openshift-kni/cnf-features-deploy/cnf-tests/testsuites/pkg/discovery"
+	"github.com/openshift-kni/cnf-features-deploy/cnf-tests/testsuites/pkg/execute"
+	"github.com/openshift-kni/cnf-features-deploy/cnf-tests/testsuites/pkg/machineconfigpool"
+	"github.com/openshift-kni/cnf-features-deploy/cnf-tests/testsuites/pkg/namespaces"
+	utilNodes "github.com/openshift-kni/cnf-features-deploy/cnf-tests/testsuites/pkg/nodes"
+	"github.com/openshift-kni/cnf-features-deploy/cnf-tests/testsuites/pkg/pods"
+	mcfgv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	mcfgScheme "github.com/openshift/machine-config-operator/pkg/generated/clientset/versioned/scheme"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	goclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const hostnameLabel = "kubernetes.io/hostname"
+const egressMCName = "qos-egress"
+const ingressMCName = "qos-ingress"
+
+type iperf3Output struct {
+	Intervals []struct {
+		Sum struct {
+			BitsPerSecond float64 `json:"bits_per_second"`
+		} `json:"sum"`
+	} `json:"intervals"`
+}
+
+var (
+	qosNodeSelector string
+	roleWorkerCNF   string
+	egressOverride  string
+	ingressOverride string
+	egressBitLimit  float64
+	ingressBitLimit float64
+	isSingleNode    bool
+	testEgress      bool
+	testIngress     bool
+)
+
+func init() {
+	roleWorkerCNF = os.Getenv("ROLE_WORKER_CNF")
+	if roleWorkerCNF != "" {
+		qosNodeSelector = fmt.Sprintf("node-role.kubernetes.io/%s=", roleWorkerCNF)
+	}
+
+	testEgress = true
+	if os.Getenv("QOS_TEST_EGRESS") == "false" {
+		testEgress = false
+	}
+
+	testIngress = true
+	if os.Getenv("QOS_TEST_INGRESS") == "false" {
+		testIngress = false
+	}
+
+	egressOverride = os.Getenv("QOS_TEST_EGRESS_OVERRIDE")
+	if egressOverride == "" {
+		egressOverride = "500000000"
+	}
+
+	ingressOverride = os.Getenv("QOS_TEST_INGRESS_OVERRIDE")
+	if ingressOverride == "" {
+		ingressOverride = "300000"
+	}
+}
+
+var _ = Describe("ovs_qos", func() {
+	gSkipReason := ""
+	execute.BeforeAll(func() {
+		err := namespaces.Create(namespaces.OVSQOSTest, client.Client)
+		Expect(err).ToNot(HaveOccurred())
+		isSingleNode, err = utilNodes.IsSingleNodeCluster()
+		Expect(err).ToNot(HaveOccurred())
+		if isSingleNode {
+			gSkipReason = "At least two nodes are required for ovs_qos tests."
+			return
+		}
+		if !discovery.Enabled() {
+			if roleWorkerCNF == "" {
+				gSkipReason = "ROLE_WORKER_CNF was not specified, can't determine nodes for ovs_qos MachineConfig without discovery."
+				return
+			}
+		}
+	})
+
+	Context("Test egress", func() {
+		var receiverNode string
+		var senderNode string
+		mcpUpdated := false
+		var err error
+		eSkipReason := ""
+		var mcpName string
+		execute.BeforeAll(func() {
```

Comment:
I think you can have a bool if the machine config exist and remove the beforeAll just have the beforeEach